### ISSUE TITLE
Feature request: support `static enum`

### ISSUE THREAD

**Issue by ashi009** [microsoft/TypeScript#34674]
## Search Terms

enum reflection, const enum, const enum literal, const enum reference, const enum export

## Suggestion

Add a new type of enum, say `static enum`, that works similar to `const enum`, but instead of inlining `const enum`, exporting enum values as constants, and emit code that refers to it.

## Use Cases

1. Long string enum values

   In case long strings are present in a `const enum`, using such enum in various places will emit code that is way too large. It would be nice to emit code that uses references to the constants instead of literals. 

2. Non-literal enum values

   More than often, we want to export a set of constants of certain types that can be passed to a function as an enum. For instance, in a base64 module, an enum representing the charset to use for encoding and decoding. In an icon module for react, an enum representing possible svg icons to display.

Some existing alternatives:

- `enum`

  `enum` emits clumsy code with runtime reflection support. However, when being used statically reflection is not needed, which makes performing dead code elimination really hard (unless closure compiler is used, which is almost infeasible when npm packages are used.)

- `export const ENUM_VALUE;`

  Practically, exporting a const works. However, there is no way to declare a type for those `ENUM_*` constants. This makes it difficult to design an API that can enforce the typing on the enum (or at least let code editor provide sensible autocompletion when a non-literal type is used.)

## Examples

icons.ts

```ts
export static enum Icons {
  SMILY_FACE = "some long svg path";
  SAD_FACE = "some long svg path";
  OBJECT = { name: "object literal", path: "some long svg path" }; // as we are just exporting consts, non literals are also ok.
}
```

app.tsx

```tsx
import { Icons } from './icons';

<Icon icon={Icons.SMILY_FACE} />
<Icon icon={Icons.OBJECT} />
```

Transpilied into:

icons.js

```js
export const Icons_SMILY_FACE = "some long svg path";
export const Icons_SAD_FACE = "some long svg path";
export const Icons_OBJECT = { name: "object literal", path: "some long svg path" };

// N.B. this is just named exports, so any current bundler will take care of unused ones.
```

app.js

```js
import { Icons_SMILY_FACE, Icons_OBJECT } from './icons';

react.createElement(Icon, { icon: Icons_SMILY_FACE });
react.createElement(Icon, { icon: Icons_OBJECT });
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
This can be accomplished today with existing syntax:
```ts
export namespace Icons {
  export const SMILY_FACE = "some long svg path";
  export const SAD_FACE = "some long svg path";
  export const OBJECT = { name: "object literal", path: "some long svg path" }; // as we are just exporting consts, non literals are also ok.
}
type Icons = (typeof Icons)[keyof typeof Icons];

function fn(x: Icons) {

}

// OK
fn(Icons.SAD_FACE);
fn(Icons.SMILY_FACE);
fn(Icons.OBJECT);
// Error
fn("blah");
```

**Comment by ashi009** [comment]
> This can be accomplished today with existing syntax:
> 
> ```ts
> export namespace Icons {
>   export const SMILY_FACE = "some long svg path";
>   export const SAD_FACE = "some long svg path";
>   export const OBJECT = { name: "object literal", path: "some long svg path" }; // as we are just exporting consts, non literals are also ok.
> }
> type Icons = (typeof Icons)[keyof typeof Icons];
> 
> function fn(x: Icons) {
> 
> }
> 
> // OK
> fn(Icons.SAD_FACE);
> fn(Icons.SMILY_FACE);
> fn(Icons.OBJECT);
> // Error
> fn("blah");
> ```

Thanks for looking into this, but this won’t work.

First, it’s not requiring that the value being passed to be one of the enum. It only assures that the type is the same. So it’s also ok to pass `{ name: "hi", path: "I just declared this"}`. If a generic typed variable is exported, say `export const OBJECT = {}`. Which will make the defined type to accept almost anything.

Second, this won’t yield a list of valid inputs for autocompletion.

**Comment by ashi009** [comment]
Please reopen this. This is not a qurstion.

**Comment by reforms** [comment]
Did I understand correctly that you need <a href="https://github.com/reforms/ts-jenum">ts-enum</a>?

**Comment by klippas-rob** [comment]
This is definitely NOT a question.
This functionality is needed - I only want to define an enum once and import it for use in multiple classes where static objects are defined.

**Comment by RyanCavanaugh** [comment]
> This is definitely NOT a question.

The question tag hasn't been on this issue for over 2 years...

**Comment by klippas-rob** [comment]
So, in two years, nothing's been done?

**Comment by RyanCavanaugh** [comment]
Correct; there are literally thousands of open suggestions and it would be infeasible (and inadvisable) to implement them all simultaneously.

**Comment by klippas-rob** [comment]
As someone new to Teams development we have found the experience nothing but frustrating:
- unless you are part of a major company that can afford $25K/year in support, support is pretty much non-existent
- unanswered questions on StackOverflow
- basic architectural flaws untouched
- dealing with simple issues takes up to 7 different people
It's been a nightmare.

So, what's the ETA on this issue? Weeks? Months? Years?
This issue is like C not supporting "includes".
Without this, there are multiple versions of the truth (repeated enum definitions scattered through the project) - fundamental, not some bling feature.

**Comment by RyanCavanaugh** [comment]
This suggestion has not garnered the quantity or quality of feedback that would lead us to prioritize it over the (again, thousands of) other open feature requests in this repo.

**Comment by klippas-rob** [comment]
Ryan,

I really do appreciate that you took the time to write - it's refreshing.
I get it - I have been long in this game and have produced many software products, and prioritisation is important.
In our deliberations on what gets implemented, sometimes what needs to be done trumps what the users clamour/vote for.

Take care.

Cheers,
Rob
,

**Comment by ashi009** [comment]
Given the latest development of typescript, 

> This can be accomplished today with existing syntax:
> 
> ```ts
> export namespace Icons {
>   export const SMILY_FACE = "some long svg path";
>   export const SAD_FACE = "some long svg path";
>   export const OBJECT = { name: "object literal", path: "some long svg path" }; // as we are just exporting consts, non literals are also ok.
> }
> type Icons = (typeof Icons)[keyof typeof Icons];
> 
> function fn(x: Icons) {
> 
> }
> 
> // OK
> fn(Icons.SAD_FACE);
> fn(Icons.SMILY_FACE);
> fn(Icons.OBJECT);
> // Error
> fn("blah");
> ```

I found that adding `as const` for each object literal will make the union type narrow enough to be useful. However, `{} as const` will still accept anything. With this hack, I think my use case is solved.

OTOH, I still prefer a dedicated way to express this, as a hack is way too error-prone.
